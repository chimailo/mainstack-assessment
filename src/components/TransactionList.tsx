import { useState } from "react";
import { ArrowDownLeft, ArrowUpRight, Check, Receipt } from "lucide-react";
import { Button } from "@/components/ui/button";
import { ChevronDown, Download } from "lucide-react";
import { FilterDrawer } from "./FilterDrawer";
import { TransactionListSkeleton } from "./TransactionListSkeleton";
import { ErrorState } from "./ErrorState";
import { useFetchTransactions } from "@/hooks/use-fetch-transactions";
import { describeDateRange, formatToCurrency } from "@/lib/utils";
import { format } from "date-fns";
import { useSetFilters } from "@/hooks/use-search-param";

export function TransactionList() {
  const [filterOpen, setFilterOpen] = useState(false);

  const { getFilters, setFilters } = useSetFilters();

  const { appliedFilters, appliedFiltersCount, hasActiveFilters } =
    getFilters();

  const {
    data: transactions,
    isLoading,
    isError,
    error,
    isFetching,
    refetch,
  } = useFetchTransactions(appliedFilters);

  const startDate = appliedFilters.startDate
    ? new Date(appliedFilters.startDate)
    : undefined;
  const endDate = appliedFilters.endDate
    ? new Date(appliedFilters.endDate)
    : undefined;

  const handleClearFilter = () => {
    setFilters({ type: null, status: null, startDate: null, endDate: null });
  };

  if (isLoading) {
    return <TransactionListSkeleton />;
  }

  return (
    <div className="mt-12">
      <div className="mb-6 flex items-center justify-between">
        <div>
          <h2 className="text-2xl font-bold flex items-center gap-2">
            {transactions?.length || 0} Transactions
            {isFetching && (
              <div className="h-2 w-2 rounded-full bg-blue-500 animate-ping" />
            )}
          </h2>
          <p className="text-sm text-muted-foreground">
            Your transactions for{" "}
            {hasActiveFilters
              ? describeDateRange(startDate, endDate)
              : "all time"}
          </p>
        </div>
        <div className="flex gap-2">
          <Button
            variant="secondary"
            size="lg"
            className="gap-2 rounded-full"
            onClick={() => setFilterOpen(true)}
          >
            Filter
            {hasActiveFilters && (
              <>
                <span className="font-medium text-sm size-5 bg-primary flex items-center justify-center">
                  {appliedFiltersCount}
                </span>
              </>
            )}
            <ChevronDown className="h-4 w-4" />
          </Button>
          <Button variant="secondary" size="lg" className="gap-2 rounded-full">
            Export list
            <Download className="h-4 w-4" />
          </Button>
        </div>
      </div>

      <FilterDrawer open={filterOpen} onOpenChange={setFilterOpen} />

      {isError ? (
        <ErrorState
          message={
            error instanceof Error
              ? error.message
              : "Failed to load transactions"
          }
          onRetry={refetch}
        />
      ) : transactions?.length === 0 ? (
        <div className="flex flex-col items-center justify-center py-16">
          <div className="mb-6 flex h-16 w-16 items-center justify-center rounded-full bg-muted">
            <Receipt className="h-8 w-8 text-muted-foreground" />
          </div>
          <h3 className="mb-2 text-xl font-semibold">
            No matching transaction found for the selected filter
          </h3>
          <p className="mb-6 text-center text-sm text-muted-foreground">
            Change your filters to see more results, or add a new product.
          </p>
          <Button variant="secondary" onClick={handleClearFilter}>
            Clear Filter
          </Button>
        </div>
      ) : (
        <div className="space-y-0">
          {(transactions || []).map((transaction) => (
            <div
              key={transaction.payment_reference}
              className="flex items-center justify-between py-4"
            >
              <div className="flex items-center gap-4">
                <div
                  className={`flex h-12 w-12 items-center justify-center rounded-full ${
                    transaction.status === "successful"
                      ? "bg-[#E3FCF2]"
                      : transaction.status === "pending"
                      ? "bg-orange-100"
                      : "bg-red-100"
                  }`}
                >
                  {transaction.status === "successful" ? (
                    <ArrowUpRight className="h-5 w-5 text-green-600" />
                  ) : (
                    <ArrowDownLeft className="h-5 w-5 text-red-600" />
                  )}
                </div>
                <div>
                  <p className="font-medium">
                    {transaction.metadata?.product_name || "N/A"}
                  </p>
                  {transaction.metadata?.name ? (
                    <p className="text-sm text-gray-500 font-medium">
                      {transaction.metadata?.name}
                    </p>
                  ) : transaction.status ? (
                    <>
                      <p
                        className={`text-sm capitalize ${
                          transaction.status === "successful"
                            ? "text-green-600"
                            : transaction.status === "pending"
                            ? "text-orange-500"
                            : "text-muted-foreground"
                        }`}
                      >
                        {transaction.status}
                      </p>
                    </>
                  ) : (
                    "â€”"
                  )}
                </div>
              </div>
              <div className="text-right">
                <p className="font-semibold">
                  {formatToCurrency(transaction.amount)}
                </p>
                <p className="text-sm text-muted-foreground">
                  {transaction.date
                    ? format(new Date(transaction.date), "MMM dd, yyyy")
                    : null}
                </p>
              </div>
            </div>
          ))}
        </div>
      )}
    </div>
  );
}
