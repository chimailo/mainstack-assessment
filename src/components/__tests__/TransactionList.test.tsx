import { render, screen, fireEvent } from "@testing-library/react";
import { describe, it, expect, vi, beforeEach } from "vitest";

// We'll mock the transaction hook and the search-param hook so we can drive
// the component through loading, error and empty states deterministically.
const mockFetch: any = {
  data: undefined,
  isLoading: true,
  isError: false,
  error: null,
  isFetching: false,
  refetch: vi.fn(),
};

const mockSetFiltersFn = vi.fn();

vi.mock("@/hooks/use-fetch-transactions", () => ({
  useFetchTransactions: () => mockFetch,
}));

vi.mock("@/hooks/use-search-param", () => ({
  useSetFilters: () => ({
    getFilters: () => ({
      appliedFilters: {},
      appliedFiltersCount: 0,
      hasActiveFilters: false,
    }),
    setFilters: mockSetFiltersFn,
  }),
}));

import { TransactionList } from "../TransactionList";

describe("TransactionList interactions", () => {
  beforeEach(() => {
    // reset to loading default before each test
    mockFetch.data = undefined;
    mockFetch.isLoading = true;
    mockFetch.isError = false;
    mockFetch.error = null;
    mockFetch.isFetching = false;
    mockFetch.refetch = vi.fn();
    mockSetFiltersFn.mockReset();
  });

  it("renders loading skeleton and matches snapshot", () => {
    mockFetch.isLoading = true;
    const { container } = render(<TransactionList />);
    expect(container).toMatchSnapshot();
  });

  it("renders error state and calls refetch when retry clicked", async () => {
    mockFetch.isLoading = false;
    mockFetch.isError = true;
    mockFetch.error = new Error("Network fail");
    const refetch = vi.fn();
    mockFetch.refetch = refetch;

    render(<TransactionList />);

    // ErrorState uses text 'Something went wrong' and 'Try again' button
    expect(screen.getByText(/Something went wrong/i)).toBeTruthy();
    const btn = screen.getByRole("button", { name: /Try again/i });
    fireEvent.click(btn);
    expect(refetch).toHaveBeenCalled();
  });

  it("renders empty state and clicking Clear Filter calls setFilters", () => {
    mockFetch.isLoading = false;
    mockFetch.isError = false;
    mockFetch.data = [];

    const { container } = render(<TransactionList />);

    expect(screen.getByText(/No matching transaction found/i)).toBeTruthy();

    const clear = screen.getByRole("button", { name: /Clear Filter/i });
    fireEvent.click(clear);
    expect(mockSetFiltersFn).toHaveBeenCalled();

    expect(container).toMatchSnapshot();
  });
});
