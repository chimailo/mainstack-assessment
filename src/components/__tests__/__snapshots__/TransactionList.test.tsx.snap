// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TransactionList interactions > renders empty state and clicking Clear Filter calls setFilters 1`] = `
<div>
  <div
    class="mt-12"
  >
    <div
      class="mb-6 flex items-center justify-between"
    >
      <div>
        <h2
          class="text-2xl font-bold flex items-center gap-2"
        >
          0
           Transactions
        </h2>
        <p
          class="text-sm text-muted-foreground"
        >
          Your transactions for
           
          all time
        </p>
      </div>
      <div
        class="flex gap-2"
      >
        <button
          class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-secondary text-secondary-foreground hover:bg-secondary/80 h-12 px-6 has-[>svg]:px-6 gap-2 rounded-full"
          data-slot="button"
        >
          Filter
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-down h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
        <button
          class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-secondary text-secondary-foreground hover:bg-secondary/80 h-12 px-6 has-[>svg]:px-6 gap-2 rounded-full"
          data-slot="button"
        >
          Export list
          <svg
            aria-hidden="true"
            class="lucide lucide-download h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 15V3"
            />
            <path
              d="M21 15v4a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2v-4"
            />
            <path
              d="m7 10 5 5 5-5"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="flex flex-col items-center justify-center py-16"
    >
      <div
        class="mb-6 flex h-16 w-16 items-center justify-center rounded-full bg-muted"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-receipt h-8 w-8 text-muted-foreground"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 2v20l2-1 2 1 2-1 2 1 2-1 2 1 2-1 2 1V2l-2 1-2-1-2 1-2-1-2 1-2-1-2 1Z"
          />
          <path
            d="M16 8h-6a2 2 0 1 0 0 4h4a2 2 0 1 1 0 4H8"
          />
          <path
            d="M12 17.5v-11"
          />
        </svg>
      </div>
      <h3
        class="mb-2 text-xl font-semibold"
      >
        No matching transaction found for the selected filter
      </h3>
      <p
        class="mb-6 text-center text-sm text-muted-foreground"
      >
        Change your filters to see more results, or add a new product.
      </p>
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-secondary text-secondary-foreground hover:bg-secondary/80 h-10 px-6 py-2 has-[>svg]:px-4"
        data-slot="button"
      >
        Clear Filter
      </button>
    </div>
  </div>
</div>
`;

exports[`TransactionList interactions > renders loading skeleton and matches snapshot 1`] = `
<div>
  <div
    class="mt-12"
  >
    <div
      class="mb-6 flex items-center justify-between"
    >
      <div>
        <div
          class="animate-pulse rounded-md bg-muted h-8 w-40"
        />
        <div
          class="animate-pulse rounded-md bg-muted mt-2 h-4 w-56"
        />
      </div>
      <div
        class="flex gap-2"
      >
        <button
          class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-secondary text-secondary-foreground hover:bg-secondary/80 h-10 px-6 py-2 has-[>svg]:px-4 gap-2 rounded-full"
          data-slot="button"
          disabled=""
        >
          Filter
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-down h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
        <button
          class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-secondary text-secondary-foreground hover:bg-secondary/80 h-10 px-6 py-2 has-[>svg]:px-4 gap-2 rounded-full"
          data-slot="button"
          disabled=""
        >
          Export list
          <svg
            aria-hidden="true"
            class="lucide lucide-download h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 15V3"
            />
            <path
              d="M21 15v4a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2v-4"
            />
            <path
              d="m7 10 5 5 5-5"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="space-y-0"
    >
      <div
        class="flex items-center justify-between border-b py-4 last:border-0"
      >
        <div
          class="flex items-center gap-4"
        >
          <div
            class="animate-pulse bg-muted h-12 w-12 rounded-full"
          />
          <div>
            <div
              class="animate-pulse rounded-md bg-muted h-5 w-32"
            />
            <div
              class="animate-pulse rounded-md bg-muted mt-1 h-4 w-24"
            />
          </div>
        </div>
        <div
          class="text-right"
        >
          <div
            class="animate-pulse rounded-md bg-muted h-5 w-20"
          />
          <div
            class="animate-pulse rounded-md bg-muted mt-1 h-4 w-16"
          />
        </div>
      </div>
      <div
        class="flex items-center justify-between border-b py-4 last:border-0"
      >
        <div
          class="flex items-center gap-4"
        >
          <div
            class="animate-pulse bg-muted h-12 w-12 rounded-full"
          />
          <div>
            <div
              class="animate-pulse rounded-md bg-muted h-5 w-32"
            />
            <div
              class="animate-pulse rounded-md bg-muted mt-1 h-4 w-24"
            />
          </div>
        </div>
        <div
          class="text-right"
        >
          <div
            class="animate-pulse rounded-md bg-muted h-5 w-20"
          />
          <div
            class="animate-pulse rounded-md bg-muted mt-1 h-4 w-16"
          />
        </div>
      </div>
      <div
        class="flex items-center justify-between border-b py-4 last:border-0"
      >
        <div
          class="flex items-center gap-4"
        >
          <div
            class="animate-pulse bg-muted h-12 w-12 rounded-full"
          />
          <div>
            <div
              class="animate-pulse rounded-md bg-muted h-5 w-32"
            />
            <div
              class="animate-pulse rounded-md bg-muted mt-1 h-4 w-24"
            />
          </div>
        </div>
        <div
          class="text-right"
        >
          <div
            class="animate-pulse rounded-md bg-muted h-5 w-20"
          />
          <div
            class="animate-pulse rounded-md bg-muted mt-1 h-4 w-16"
          />
        </div>
      </div>
      <div
        class="flex items-center justify-between border-b py-4 last:border-0"
      >
        <div
          class="flex items-center gap-4"
        >
          <div
            class="animate-pulse bg-muted h-12 w-12 rounded-full"
          />
          <div>
            <div
              class="animate-pulse rounded-md bg-muted h-5 w-32"
            />
            <div
              class="animate-pulse rounded-md bg-muted mt-1 h-4 w-24"
            />
          </div>
        </div>
        <div
          class="text-right"
        >
          <div
            class="animate-pulse rounded-md bg-muted h-5 w-20"
          />
          <div
            class="animate-pulse rounded-md bg-muted mt-1 h-4 w-16"
          />
        </div>
      </div>
      <div
        class="flex items-center justify-between border-b py-4 last:border-0"
      >
        <div
          class="flex items-center gap-4"
        >
          <div
            class="animate-pulse bg-muted h-12 w-12 rounded-full"
          />
          <div>
            <div
              class="animate-pulse rounded-md bg-muted h-5 w-32"
            />
            <div
              class="animate-pulse rounded-md bg-muted mt-1 h-4 w-24"
            />
          </div>
        </div>
        <div
          class="text-right"
        >
          <div
            class="animate-pulse rounded-md bg-muted h-5 w-20"
          />
          <div
            class="animate-pulse rounded-md bg-muted mt-1 h-4 w-16"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
